syntax = "proto3";

package product;

service ProductService {
    rpc CreateProduct (ProductRequest) returns (ProductResponse);
    rpc GetProduct (ProductId) returns (ProductResponse);
    rpc ListProducts (Empty) returns (ProductList);
    rpc UpdateProduct (ProductUpdateRequest) returns (ProductResponse);
    rpc DeleteProduct (ProductId) returns (DeleteResponse);
}

message Empty {}

message ProductId {
    string id = 1;
}

message ProductRequest {
    string name = 1;
    string description = 2;
    double price = 3;
}

message ProductUpdateRequest {
  string id = 1;
  string name = 2;
  string description = 3;
  double price = 4;
}

message ProductResponse {
    string id = 1;
    string name = 2;
    string description = 3;
    double price = 4;
}

message ProductList {
    repeated ProductResponse products = 1;
}

message DeleteResponse {
    string id = 1;
}
